### PR TITLE
Fix flaky BigQuery tests using snapshot tables

### LINKED ISSUE #18266: Broken `testEmptyProjection` in BigQuery connector
```
Error:  io.trino.plugin.bigquery.TestBigQueryAvroConnectorTest.testEmptyProjection[SNAPSHOT, CREATE SNAPSHOT TABLE %s CLONE tpch.region, DROP SNAPSHOT TABLE %s](5)  Time elapsed: 0.456 s  <<< FAILURE!
com.google.cloud.bigquery.BigQueryException: Cannot have more than 1000 active clones and snapshots of a table. Please use a deep copy, or remove some active clones or snapshots.
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.translate(HttpBigQueryRpc.java:115)
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.queryRpc(HttpBigQueryRpc.java:720)
	at com.google.cloud.bigquery.BigQueryImpl$35.call(BigQueryImpl.java:1322)
	at com.google.cloud.bigquery.BigQueryImpl$35.call(BigQueryImpl.java:1319)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at com.google.cloud.bigquery.BigQueryRetryHelper.run(BigQueryRetryHelper.java:86)
	at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:49)
	at com.google.cloud.bigquery.BigQueryImpl.queryRpc(BigQueryImpl.java:1318)
	at com.google.cloud.bigquery.BigQueryImpl.query(BigQueryImpl.java:1306)
	at io.trino.plugin.bigquery.BigQueryQueryRunner$BigQuerySqlExecutor.executeQuery(BigQueryQueryRunner.java:149)
	at io.trino.plugin.bigquery.BigQueryQueryRunner$BigQuerySqlExecutor.execute(BigQueryQueryRunner.java:143)
	at io.trino.plugin.bigquery.BaseBigQueryConnectorTest.onBigQuery(BaseBigQueryConnectorTest.java:995)
	at io.trino.plugin.bigquery.BaseBigQueryConnectorTest.testEmptyProjection(BaseBigQueryConnectorTest.java:227)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
POST https://www.googleapis.com/bigquery/v2/projects/sep-bq-cicd/queries
{
  "code": 400,
  "errors": [
    {
      "domain": "global",
      "message": "Cannot have more than 1000 active clones and snapshots of a table. Please use a deep copy, or remove some active clones or snapshots.",
      "reason": "invalid"
    }
  ],
  "message": "Cannot have more than 1000 active clones and snapshots of a table. Please use a deep copy, or remove some active clones or snapshots.",
  "status": "INVALID_ARGUMENT"
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$3.interceptResponse(AbstractGoogleClientRequest.java:466)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:552)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:493)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:603)
	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.queryRpc(HttpBigQueryRpc.java:718)
	... 24 more
```

https://github.com/trinodb/trino/actions/runs/5525099575/jobs/10080465642?pr=17667

### PR DESCRIPTION
**Pull request by hashhar**
## Description

BigQuery has limits on how many snapshots or clones can exist for a single table. It seems BigQuery is miscounting though because we occasionally see errors like below on CI even though all snapshots get dropped by their respective tests:

    Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
    POST https://www.googleapis.com/bigquery/v2/projects/sep-bq-cicd/queries
    {
      "code": 400,
      "errors": [
        {
          "domain": "global",
          "message": "Cannot have more than 1000 active clones and snapshots of a table. Please use a deep copy, or remove some active clones or snapshots.",
          "reason": "invalid"
        }
      ],
      "message": "Cannot have more than 1000 active clones and snapshots of a table. Please use a deep copy, or remove some active clones or snapshots.",
      "status": "INVALID_ARGUMENT"
    }

A workaround is to use a different table as the source of the snapshot everytime which is what this change does.

Fixes #18266

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 1]
```diff
@@ -240,11 +247,11 @@ public void testEmptyProjection(TableDefinition.Type tableType, String createSql
...
-                {TABLE, "CREATE TABLE %s AS SELECT * FROM tpch.region", "DROP TABLE %s"},
-                {VIEW, "CREATE VIEW %s AS SELECT * FROM tpch.region", "DROP VIEW %s"},
-                {MATERIALIZED_VIEW, "CREATE MATERIALIZED VIEW %s AS SELECT * FROM tpch.region", "DROP MATERIALIZED VIEW %s"},
-                {EXTERNAL, "CREATE EXTERNAL TABLE %s OPTIONS (format = 'CSV', uris = ['gs://" + gcpStorageBucket + "/tpch/tiny/region.csv'])", "DROP EXTERNAL TABLE %s"},
-                {SNAPSHOT, "CREATE SNAPSHOT TABLE %s CLONE tpch.region", "DROP SNAPSHOT TABLE %s"},
+                {TABLE, "CREATE TABLE %s %s", "AS SELECT * FROM tpch.region", "DROP TABLE %s"},
+                {VIEW, "CREATE VIEW %s %s", "SELECT * FROM tpch.region", "DROP VIEW %s"},
+                {MATERIALIZED_VIEW, "CREATE MATERIALIZED VIEW %s %s", "AS SELECT * FROM tpch.region", "DROP MATERIALIZED VIEW %s"},
+                {EXTERNAL, "CREATE EXTERNAL TABLE %s %s", "OPTIONS (format = 'CSV', uris = ['gs://" + gcpStorageBucket + "/tpch/tiny/region.csv'])", "DROP EXTERNAL TABLE %s"},
+                {SNAPSHOT, "CREATE SNAPSHOT TABLE %s %s", "CLONE %s", "DROP SNAPSHOT TABLE %s"},
```
Comment (hashhar) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:254]: couldn't think of a cleaner alternative than parameterizing the "table definition" as well. Suggestions welcome.

Comment (hashhar) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:254]: Fair point.

**Review by wendigo — APPROVED** [review]

**Comment by hashhar** [comment]
AC @ebyhr. PTAL again.

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 2]
```diff
@@ -219,40 +208,71 @@ public void testCreateTableIfNotExists()
...
+        // For snapshot table test we use a different source table everytime
+        String regionCopy = TEST_SCHEMA + ".region_" + randomNameSuffix();
+        try {
+            onBigQuery("CREATE TABLE " + regionCopy + " AS SELECT * FROM tpch.region");
+            testEmptyProjection(
+                    snapshotTableName -> onBigQuery("CREATE SNAPSHOT TABLE " + snapshotTableName + " CLONE " + regionCopy),
+                    snapshotTableName -> onBigQuery("DROP SNAPSHOT TABLE " + snapshotTableName));
+        }
+        finally {
+            onBigQuery("DROP TABLE " + regionCopy);
```
[the remark this answers is not in the document]
Comment (hashhar) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:256]: Fixed. Thanks.

**Review by ssheikin — APPROVED** [review]

**Pull request merged**